### PR TITLE
Add api service into column

### PR DESCRIPTION
**Pull request by ushergod**
This is a stractch patch for column api by using flask api.
Three routes:
    1. GET: runs/id (return a run)
    2. POST: runs (trigger a new run)
    3. GET: runs (return runs list)

I would update the route as we update the swagger file

TODO:
1. Add exception
2. Add log
3. Add config for api
4. Add client payload validation
5. Add persistent backed
6. Add start point in the package install script

Welcome comments!

### CODE REVIEW

File: column/api/manager/run_manager.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        if 'state' in column_opts:
+            del (column_opts['state'])
+        if 'id' in column_opts:
+            del (column_opts['id'])
+        return column_opts
+
+    def _run_playbook(self, run):
+        playbook = run['playbook_path']
+        column_opts = self._build_opts(run)
+        self.column_manager.run_playbook(run['playbook_path'],
```
Comment (zw2326) [inline column/api/manager/run_manager.py]: run['playbook_path'] -> playbook? Otherwise it seems not to be used...

File: column/api/manager/run_manager.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    def __init__(self):
+        self.column_manager = column.APIRunner()
+
+    def _build_opts(self, opts):
+        column_opts = copy.deepcopy(opts)
+        # state attribute is not useful to column runner
+        if 'state' in column_opts:
+            del (column_opts['state'])
+        if 'id' in column_opts:
```
Comment (zw2326) [inline column/api/manager/run_manager.py]: Worth a comment like above?

File: column/api/backend/cache/local_mem.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+from collections import deque  # noqa
+import threading
+
+import store
+
+
+# Shared dictionaries for LocalMemoryCache
+_stores = {}
```
Comment (zw2326) [inline column/api/backend/cache/local_mem.py]: Just a thought - move these variables to LocalMemoryStore or store.Store as class static members?

File: column/api/backend/cache/local_mem.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+class LocalMemoryStore(store.Store):
+
+    def __init__(self, name):
+        self._store = _stores.setdefault(name, {})
+        self._key_queue = _key_queues.setdefault(name, deque([]))
+        lock = _locks.setdefault(name, threading.Lock())
+        super(LocalMemoryStore, self).__init__(name, lock)
+
+    def _has_key(self, key):
+        if key in self._store:
```
Comment (zw2326) [inline column/api/backend/cache/local_mem.py]: Or:
def _has_key(self, key):
    return key in self._store

File: column/api/backend/cache/local_mem.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            return True
+        return False
+
+    def _add(self, key, value):
+        self._save(key, value)
+        self._key_queue.append(key)
+
+    def _del(self, del_key):
```
Comment (zw2326) [inline column/api/backend/cache/local_mem.py]: "del_key" -> "key" for consistency with its parent class (or the other way round, whichever easier)

File: column/api/backend/cache/local_mem.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        del self._store[del_key]
+
+    def _save(self, key, value):
+        self._store[key] = value
+
+    def _retrieve(self, key):
+        return self._store[key]
+
+    def _is_full(self):
+        if len(self._key_queue) >= store.MAX_STORE_SIZE:
```
Comment (zw2326) [inline column/api/backend/cache/local_mem.py]: Or:
return len(self._key_queue) >= store.MAX_STORE_SIZE

File: column/api/backend/cache/local_mem.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            return True
+
+    def _evict(self):
+        del_key = self._key_queue.popleft()
+        del self._store[del_key]
```
Comment (zw2326) [inline column/api/backend/cache/local_mem.py]: Maybe call self._del() to avoid duplicated code?

File: column/api/backend/cache/store.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            if self._has_key(key):
+                self._del(key)
+                LOG.debug("Deleted key %s from store %s", key, self.name)
+                return True
+            return False
+
+    def get(self, key):
+        with self.lock:
+            if self._has_key(key):
+                LOG.debug("Get key %s from store %s", key, self.name)
```
Comment (zw2326) [inline column/api/backend/cache/store.py]: "Get" -> "Got" for consistency?

File: column/api/backend/cache/store.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+    def keys(self):
+        with self.lock:
+            return self._keys()
+
+    def set(self, key, value):
+        with self.lock:
+            if self._has_key(key):
+                self._save(key, value)
+                LOG.debug("Updating key %s from store %s", key, self.name)
```
Comment (zw2326) [inline column/api/backend/cache/store.py]: "Updating" -> "Updated" for consistency?

File: column/api/manager/run_manager.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        column_opts = self._build_opts(run)
+        self.column_manager.run_playbook(run['playbook_path'],
+                                         **column_opts)
+        run['state'] = State.COMPLETED
+
+    def create_run(self, run):
+        LOG.debug('Triggering a new run with %s', run)
+        run['state'] = State.RUNNING
+        t = threading.Thread(target=self._run_playbook,
+                             args=[run])
```
Comment (zw2326) [inline column/api/manager/run_manager.py]: You sure it is "args=[run]", not "args=(run,)"?

File: column/api/run.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
+from flask import Flask
+from flask_restful import Api
+from column.api.controller.run_controller import Run
```
Comment (zw2326) [inline column/api/run.py]: I don't know what your coding standard is, but maybe combine these two imports for simplicity?

File: column/api/controller/run_controller.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
+import uuid
+
+from flask import request
+from flask_restful import Resource
+
+from column.api import backend
```
Comment (zw2326) [inline column/api/controller/run_controller.py]: I don't know what your coding standard is, but maybe combine these two imports for simplicity?

File: column/api/backend/cache/local_mem.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+import store
+
+
+# Shared dictionaries for LocalMemoryCache
+_stores = {}
+_key_queues = {}
+_locks = {}
+
+MAX_STORE_SIZE = 50
```
Comment (zw2326) [inline column/api/backend/cache/local_mem.py]: Why is there a MAX_STORE_SIZE in both column/api/backend/cache/local_mem.py and column/api/backend/cache/store.py?

File: column/api/manager/run_manager.py (hunk 1 above)
Comment (ushergod) [inline column/api/manager/run_manager.py]: good catch

File: column/api/controller/run_controller.py (hunk 12 above)
Comment (ushergod) [inline column/api/controller/run_controller.py]: E401 multiple imports on one line

File: column/api/run.py (hunk 11 above)
Comment (ushergod) [inline column/api/run.py]: E401 multiple imports on one line

File: column/api/backend/cache/local_mem.py (hunk 13 above)
Comment (ushergod) [inline column/api/backend/cache/local_mem.py]: will update it

File: column/api/backend/cache/local_mem.py (hunk 5 above)
Comment (ushergod) [inline column/api/backend/cache/local_mem.py]: will update it

File: column/api/backend/cache/local_mem.py (hunk 4 above)
Comment (ushergod) [inline column/api/backend/cache/local_mem.py]: yes, that would be easier

File: column/api/backend/cache/local_mem.py (hunk 6 above)
Comment (ushergod) [inline column/api/backend/cache/local_mem.py]: sure

File: column/api/backend/cache/local_mem.py (hunk 7 above)
Comment (ushergod) [inline column/api/backend/cache/local_mem.py]: sure

File: column/api/backend/cache/store.py (hunk 8 above)
Comment (ushergod) [inline column/api/backend/cache/store.py]: sure

File: column/api/backend/cache/store.py (hunk 9 above)
Comment (ushergod) [inline column/api/backend/cache/store.py]: sure

File: .gitignore [hunk 14]
```diff
@@ -1,3 +1,6 @@
 build/
 *.egg-info
 dist/
+*.pyc
```
Comment (ericwb) [inline .gitignore]: These are unrelated and better in a separate commit.

File: column/api/backend/cache/local_mem.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
+import copy
+import logging
+import threading
+from collections import deque  # noqa
```
Comment (ericwb) [inline column/api/backend/cache/local_mem.py]: pylint: Not alpha sorted

File: setup.py [hunk 16]
```diff
@@ -30,6 +30,7 @@
     author='VMware',
     author_email='openstack@vmware.com',
     description='A thin wrapper on top of ansible with a stable API.',
+    data_files=[('/etc', ['etc/column.conf'])],
```
Comment (ericwb) [inline setup.py]: better to have in its own directory as etc/column/column.conf

File: requirements.txt [hunk 17]
```diff
@@ -1,2 +1,3 @@
 ansible>=2.2.0.0
 six
+flask
```
Comment (ericwb) [inline requirements.txt]: sort

File: etc/column.conf [hunk 18]
```diff
@@ -0,0 +1,10 @@
+[DEFAULT]
+# Log file
+log_file = /var/log/column.log
+
+# Log info level
+log_level = DEBUG
+
+[API]
+server = 127.0.0.1
+port = 5000
```
Comment (ericwb) [inline etc/column.conf]: Don't use 5000.  It's keystone's port and probably some others.  Find a better unused port.

File: column/api/objects/__init__.py [hunk 19]
```diff
@@ -0,0 +1,3 @@
+class State(object):
```
Comment (ericwb) [inline column/api/objects/__init__.py]: Is a class necessary for a basic enum?

File: column/api/manager/run_manager.py [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+class ColumnManager(object):
+
+    def __init__(self):
+        self.column_manager = column.APIRunner()
+
+    def _build_opts(self, opts):
+        column_opts = copy.deepcopy(opts)
+        # state attribute is not useful to column runner
+        if 'state' in column_opts:
+            del (column_opts['state'])
```
Comment (ericwb) [inline column/api/manager/run_manager.py]: remove ( ) around column_opts

File: column/api/manager/run_manager.py [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+        # state attribute is not useful to column runner
+        if 'id' in column_opts:
+            del (column_opts['id'])
```
Comment (ericwb) [inline column/api/manager/run_manager.py]: remove ( ) around column_opts

File: column/api/manager/__init__.py [hunk 22]
```diff
@@ -0,0 +1,6 @@
+import run_manager
```
Comment (ericwb) [inline column/api/manager/__init__.py]: improper import. use full namespace

File: column/api/controller/run_controller.py (hunk 12 above)
Comment (ericwb) [inline column/api/controller/run_controller.py]: Don't add multiple imports on the same line please.  Makes it a mess

File: column/api/backend/cache/local_mem.py [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+
+import store
```
Comment (ericwb) [inline column/api/backend/cache/local_mem.py]: Improper import. Use full path

File: .gitignore (hunk 14 above)
Comment (ushergod) [inline .gitignore]: sure

File: etc/column.conf (hunk 18 above)
Comment (ushergod) [inline etc/column.conf]: plan to use 48620

File: column/api/objects/__init__.py (hunk 19 above)
Comment (ushergod) [inline column/api/objects/__init__.py]: I think it is a common practice for enum.

File: column/api/backend/cache/local_mem.py (hunk 15 above)
Comment (ushergod) [inline column/api/backend/cache/local_mem.py]: sure

File: column/api/backend/cache/local_mem.py (hunk 23 above)
Comment (ushergod) [inline column/api/backend/cache/local_mem.py]: sure

File: setup.py (hunk 16 above)
Comment (ushergod) [inline setup.py]: sure

File: requirements.txt (hunk 17 above)
Comment (ushergod) [inline requirements.txt]: sure

File: column/api/manager/run_manager.py (hunk 20 above)
Comment (ushergod) [inline column/api/manager/run_manager.py]: sure

File: column/api/manager/run_manager.py (hunk 21 above)
Comment (ushergod) [inline column/api/manager/run_manager.py]: sure

File: column/api/manager/__init__.py (hunk 22 above)
Comment (ushergod) [inline column/api/manager/__init__.py]: sure

File: column/api/backend/cache/store.py [hunk 24]
```diff
@@ -0,0 +1,78 @@
+import logging
+
+LOG = logging.getLogger(__name__)
+
+MAX_STORE_SIZE = 50
+
+
```
Comment (ericwb) [inline column/api/backend/cache/store.py]: Add docstrings

File: column/api/backend/cache/local_mem.py [hunk 25]
```diff
@@ -0,0 +1,70 @@
+import copy
+from collections import deque  # noqa
+import logging
+import threading
+
```
Comment (ericwb) [inline column/api/backend/cache/local_mem.py]: Add docstrings

File: column/api/backend/cache/__init__.py [hunk 26]
```diff
@@ -0,0 +1,21 @@
+import local_mem
+
+
+class LocalMemoryCache(object):
```
Comment (ericwb) [inline column/api/backend/cache/__init__.py]: Why is this class in an __init__ file?

File: column/api/controller/run_controller.py [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+from column.api import manager
+
+LOG = logging.getLogger(__name__)
+
+
+class Run(Resource):
```
Comment (ericwb) [inline column/api/controller/run_controller.py]: Add docstrings

File: column/api/controller/run_controller.py [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+
+    def __init__(self):
+        self.backend_api = backend.get_api()
+
+    def get(self, id):
+        return self.backend_api.get_run(id)
+
+
+class RunList(Resource):
```
Comment (ericwb) [inline column/api/controller/run_controller.py]: Add docstrings

File: column/api/manager/run_manager.py [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+
+import column
+from column.api.objects import State
+from column.plugins.callback.progress import AnsibleTrackProgress
+
+
+LOG = logging.getLogger(__name__)
+
+
+class ColumnManager(object):
```
Comment (ericwb) [inline column/api/manager/run_manager.py]: Add docstrings

File: column/api/manager/run_manager.py [hunk 30]
```diff
@@ -0,0 +1,43 @@
+import copy
+import logging
+import threading
+
+import column
+from column.api.objects import State
+from column.plugins.callback.progress import AnsibleTrackProgress
```
Comment (ericwb) [inline column/api/manager/run_manager.py]: Don't import classes, along modules

File: column/api/manager/run_manager.py [hunk 31]
```diff
@@ -0,0 +1,43 @@
+import copy
+import logging
+import threading
+
+import column
+from column.api.objects import State
```
Comment (ericwb) [inline column/api/manager/run_manager.py]: Don't import classes, along modules

File: column/api/run.py [hunk 32]
```diff
@@ -0,0 +1,28 @@
+import logging
+
+from flask import Flask
```
Comment (ericwb) [inline column/api/run.py]: Why Flask and not WSGI?  With wsgi we'll be able to have more options in how we stand up the service (apache2 or uwsgi, etc).  Also Flask’s built-in server is not suitable for production as it doesn’t scale well and by default serves only one request at a time.

File: column/api/run.py [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+from flask_restful import Api
+
+from column import cfg
+from column.api.controller.run_controller import Run
```
Comment (ericwb) [inline column/api/run.py]: Do not import classes, only modules

File: column/api/run.py [hunk 34]
```diff
@@ -0,0 +1,28 @@
+import logging
+
+from flask import Flask
+from flask_restful import Api
```
Comment (ericwb) [inline column/api/run.py]: Do not import classes, only modules

File: column/api/run.py [hunk 35]
```diff
@@ -0,0 +1,28 @@
...
+
+from column import cfg
+from column.api.controller.run_controller import Run
+from column.api.controller.run_controller import RunList
```
Comment (ericwb) [inline column/api/run.py]: Do not import classes, only modules

File: column/plugins/callback/progress.py [hunk 36]
```diff
@@ -0,0 +1,32 @@
+import logging
+
+import column
+
+LOG = logging.getLogger(__name__)
+
+
+class AnsibleTrackProgress(column.AnsibleCallback):
```
Comment (ericwb) [inline column/plugins/callback/progress.py]: Add docstrings

File: etc/column/column.conf [hunk 37]
```diff
@@ -0,0 +1,10 @@
+[DEFAULT]
+# Log file
+log_file = /var/log/column.log
+
+# Log info level
+log_level = DEBUG
+
+[API]
+server = 0.0.0.0
```
Comment (ericwb) [inline etc/column/column.conf]: Don't listen on address 0.0.0.0 (all interfaces) by default. This can be a potential security issue if someone doesn't realize the service is bound to an insecure interface.

File: tests/api/controllers/test_run.py [hunk 38]
```diff
@@ -0,0 +1,59 @@
...
+import time
+
+from tests.api.controllers import APITest
+
+
+class TestRun(APITest):
+
+    def _wait_for_run_complete(self, id):
+        counter = 0
+        while 1:
```
Comment (ericwb) [inline tests/api/controllers/test_run.py]: use True not 1

File: column/__init__.py [hunk 39]
```diff
@@ -7,10 +7,16 @@
...
 
 __all__ = [
     'APIRunner', 'Runner', 'SubprocessRunner', 'AnsibleCallback'
 ]
 
 
 __version__ = '0.3.6'
+
+cfg = ConfigParser.ConfigParser(allow_no_value=True)
+cfg.read(DEFAULT_CONF_FILE)
```
Comment (kars7e) [inline column/__init__.py]: What happens if this file does not exist? This will be executed globally for column, even when column is used as library and not as a service.

File: column/api/backend/__init__.py [hunk 40]
```diff
@@ -0,0 +1,6 @@
+import cache
+
+
+def get_api():
```
Comment (kars7e) [inline column/api/backend/__init__.py]: this name is very confusing. Why is this named get_api? get_cache would be better, or even "get_store", since in the end what we need here is some sort of store to keep our state.

File: column/api/run.py (hunk 32 above)
Comment (kars7e) [inline column/api/run.py]: Flask is compatible with WSGI, and can be deployed with apache2 and uwsgi.
http://flask.pocoo.org/docs/0.12/deploying/.
Built-in server is definitely for development and testing only, we need a dockerfile with apache or nginx instead.

File: etc/column/column.conf (hunk 37 above)
Comment (kars7e) [inline etc/column/column.conf]: I think this is fine as a default value. It's deployer's duty to figure out the actual interface service should be bound to, we can't make this decision for him/her. At the same time, we need something here for testing/development.

File: tests/api/controllers/test_run.py [hunk 41]
```diff
@@ -0,0 +1,59 @@
...
+            if res_dict['state'] != 'RUNNING':
+                break
+            time.sleep(self.wait_interval)
+            if counter < self.counter:
+                continue
+        self.assertNotEqual('RUNNING', res_dict['state'])
+
+    def test_get_empty_runlist(self):
+        response = self.app.get('/runs')
+        self.assertEqual(200, response.status_code)
```
Comment (ericwb) [inline tests/api/controllers/test_run.py]: Instead of 200, just use constants from http_client

from six.moves import http_client

http_client.OK = 200

File: tests/api/controllers/test_run.py [hunk 42]
```diff
@@ -0,0 +1,59 @@
...
+
+    def test_trigger_run(self):
+        pb = 'tests/fixtures/playbooks/hello_world.yml'
+        response = self.app.post('/runs', data=json.dumps(
+                                            dict(playbook_path=pb,
+                                            inventory_file='localhosti,',
+                                            connection='local')),
+                                            content_type='application/json')
+        res_dict = json.loads(response.data)
+        self.assertEqual(200, response.status_code)
```
Comment (ericwb) [inline tests/api/controllers/test_run.py]: Instead of 200, just use constants from http_client

from six.moves import http_client

http_client.OK = 200

File: etc/column/column.conf (hunk 37 above)
Comment (ericwb) [inline etc/column/column.conf]: It's a security best practice.  Using 127.0.0.1 is much better.  And rarely if ever do deployers do the due diligence or even aware of the consequences.

File: column/plugins/callback/progress.py [hunk 43]
```diff
@@ -0,0 +1,32 @@
...
+    def __init__(self, *args, **kwargs):
+        super(self.__class__, self).__init__(*args, **kwargs)
+        self.total_plays = 0
+        self.finished_plays = 0
+        self.playbook = None
+        self.progress = 0
+
+    def run_on_playbook_start(self, playbook):
+        self.playbook = playbook._file_name
```
Comment (ericwb) [inline column/plugins/callback/progress.py]: This reminds me that we need to refactor the callbacks so that ansible attributes don't need to be accessed like this example.

File: column/api/backend/cache/__init__.py [hunk 44]
```diff
@@ -0,0 +1,21 @@
...
+        run = self.runs_store.get_run(run_id)
+        return run
+
+    def create_run(self, run_id, run):
+        return self.runs_store.add_run(run_id, run)
+
+    def list_runs(self):
+        run_ids = self.runs_store.keys()
+        runs = []
+        for run_id in run_ids:
```
Comment (ericwb) [inline column/api/backend/cache/__init__.py]: Just use list comprehension.

File: column/api/run.py (hunk 32 above)
Comment (ericwb) [inline column/api/run.py]: Why start with something that doesn't scale?

Comment (kars7e) [inline column/api/run.py]: I agree python is not the best choice, but we were kind of stuck with it based on the fact ansible is using python. I would prefer Go myself :)

File: etc/column/column.conf (hunk 37 above)
Comment (kars7e) [inline etc/column/column.conf]: I wish you included that in your first comment, my impression was that you wanted to not have any default at all :-) I agree, 127.0.0.1 sounds much better!

**Comment by kars7e** [comment]
This is good point Eric, I agree the in-memory thing is not a way to go moving forward, also keeping the state with the app is not in par with 12 factor apps. The goal here is to have some generic store interface and then use the proper backend based on the use case. Then in-memory thing will be still good for development and initial testing, however something external (like etcd for example) should be added eventually.

File: column/api/manager/run_manager.py (hunk 29 above)
Comment (ushergod) [inline column/api/manager/run_manager.py]: sure

File: column/api/controller/run_controller.py (hunk 28 above)
Comment (ushergod) [inline column/api/controller/run_controller.py]: sure

File: column/api/controller/run_controller.py (hunk 27 above)
Comment (ushergod) [inline column/api/controller/run_controller.py]: sure

File: column/api/backend/cache/local_mem.py (hunk 25 above)
Comment (ushergod) [inline column/api/backend/cache/local_mem.py]: sure

File: column/api/backend/cache/store.py (hunk 24 above)
Comment (ushergod) [inline column/api/backend/cache/store.py]: sure

File: column/api/manager/run_manager.py (hunk 31 above)
Comment (ushergod) [inline column/api/manager/run_manager.py]: sure

File: column/api/run.py (hunk 33 above)
Comment (ushergod) [inline column/api/run.py]: sure

File: column/api/run.py (hunk 34 above)
Comment (ushergod) [inline column/api/run.py]: sure

File: column/api/run.py (hunk 35 above)
Comment (ushergod) [inline column/api/run.py]: sure

File: column/plugins/callback/progress.py (hunk 36 above)
Comment (ushergod) [inline column/plugins/callback/progress.py]: sure

File: etc/column/column.conf (hunk 37 above)
Comment (ushergod) [inline etc/column/column.conf]: 127.0.0.1 is fine, but when we build the docker image, it should be 0.0.0.0.

File: tests/api/controllers/test_run.py (hunk 38 above)
Comment (ushergod) [inline tests/api/controllers/test_run.py]: sure

File: column/__init__.py (hunk 39 above)
Comment (ushergod) [inline column/__init__.py]: according to the doc, https://docs.python.org/2/library/configparser.html#ConfigParser.RawConfigParser.read

If none of the named files exist, the ConfigParser instance will contain an empty dataset.
So i think it is okay now

File: column/api/backend/__init__.py (hunk 40 above)
Comment (ushergod) [inline column/api/backend/__init__.py]: sure

File: tests/api/controllers/test_run.py (hunk 41 above)
Comment (ushergod) [inline tests/api/controllers/test_run.py]: sure

File: tests/api/controllers/test_run.py (hunk 42 above)
Comment (ushergod) [inline tests/api/controllers/test_run.py]: sure

File: column/api/backend/cache/__init__.py (hunk 44 above)
Comment (ushergod) [inline column/api/backend/cache/__init__.py]: sure

File: tests/api/controllers/test_run.py [hunk 45]
```diff
@@ -0,0 +1,60 @@
+import json
+from six.moves import http_client
```
Comment (ericwb) [inline tests/api/controllers/test_run.py]: The six module is 3rd party, so should be grouped with base modules.

File: tests/api/controllers/test_run.py [hunk 46]
```diff
@@ -0,0 +1,60 @@
...
+import time
+
+from tests.api.controllers import APITest
```
Comment (ericwb) [inline tests/api/controllers/test_run.py]: Don't import classes.

File: tests/api/controllers/__init__.py [hunk 47]
```diff
@@ -0,0 +1,14 @@
+from testtools import TestCase  # noqa
```
Comment (ericwb) [inline tests/api/controllers/__init__.py]: don't import classes

File: setup.py [hunk 48]
```diff
@@ -30,8 +30,9 @@
     author='VMware',
     author_email='openstack@vmware.com',
```
Comment (ericwb) [inline setup.py]: Shouldn't use openstack@vmware.com.

File: column/api/controller/run_controller.py [hunk 49]
```diff
@@ -0,0 +1,37 @@
...
+from flask_restful import Resource
+
+from column.api import backend
+from column.api import manager
+
+LOG = logging.getLogger(__name__)
+
+
+class Run(Resource):
+    """column api run resource class """
```
Comment (ericwb) [inline column/api/controller/run_controller.py]: These comments don't state anything other than the obvious.

File: column/api/controller/run_controller.py [hunk 50]
```diff
@@ -0,0 +1,37 @@
...
+    def __init__(self):
+        self.backend_store = backend.get_store()
+
+    def get(self, id):
+        return self.backend_store.get_run(id)
+
+
+class RunList(Resource):
+    """column api Runlist resource list class """
```
Comment (ericwb) [inline column/api/controller/run_controller.py]: These comments don't state anything other than the obvious.

File: column/api/backend/cache/store.py [hunk 51]
```diff
@@ -0,0 +1,79 @@
...
+class Store(object):
+    """Backend store base class """
```
Comment (ericwb) [inline column/api/backend/cache/store.py]: These comments don't state anything other than the obvious.

File: column/api/backend/cache/local_mem.py [hunk 52]
```diff
@@ -0,0 +1,70 @@
...
+        del_key = self._key_queue.popleft()
+        self._del[del_key]
+        return del_key
+
+    def _keys(self):
+        return self._store.keys()
+
+
+class RunMemoryStore(LocalMemoryStore):
+    """Column api service in-memory backend interface class """
```
Comment (ericwb) [inline column/api/backend/cache/local_mem.py]: These comments don't state anything other than the obvious.

File: column/api/backend/cache/local_mem.py [hunk 53]
```diff
@@ -0,0 +1,70 @@
...
+# Shared dictionaries for LocalMemoryCache
+_stores = {}
+_key_queues = {}
+_locks = {}
+
+LOG = logging.getLogger(__name__)
+
+
+class LocalMemoryStore(store.Store):
+    """In memory backend base class """
```
Comment (ericwb) [inline column/api/backend/cache/local_mem.py]: These comments don't state anything other than the obvious.

File: column/api/objects/__init__.py [hunk 54]
```diff
@@ -0,0 +1,3 @@
...
+    RUNNING = "RUNNING"
```
Comment (tracyajones) [inline column/api/objects/__init__.py]: what about failed?

File: column/api/manager/run_manager.py [hunk 55]
```diff
@@ -0,0 +1,43 @@
...
+            del column_opts['id']
+        return column_opts
+
+    def _run_playbook(self, run):
+        column_opts = self._build_opts(run)
+        progress_callback = progress.AnsibleTrackProgress()
+        self.column_manager.add_callback(progress_callback)
+        self.column_manager.run_playbook(run['playbook_path'],
+                                         **column_opts)
+        run['state'] = State.COMPLETED
```
Comment (tracyajones) [inline column/api/manager/run_manager.py]: when a play failes how do we access the params (see ansible_util._parse_results

File: column/api/objects/__init__.py (hunk 54 above)
Comment (ushergod) [inline column/api/objects/__init__.py]: In the api side currently, it can not get the state of ansible run. It will involve changes of column library to provide the interface to get the  run state. 

i would like to put it into the future task

Comment (tracyajones) [inline column/api/objects/__init__.py]: im ok with that but we do need to be able to access the results as in my comment below

File: column/api/manager/run_manager.py (hunk 55 above)
Comment (ushergod) [inline column/api/manager/run_manager.py]: Yes, I would keep updating the patch for supporting this.

**Comment by ushergod** [comment]
Updates for apache mod_wsgi, apache will not reload the application when we set 'WSGIDaemonProcess'.  Since current backend implementation is in-memory, so setting more than one process in apache conf will fail the application. One process with multiple thread is running fine.
